Sources/SwiftLexicalScopes/Scope.swift

```diff
@@ -0,0 +1,139 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import SwiftSyntax
+
+extension SyntaxProtocol {
+  /// Scope at the syntax node. Could be inherited from parent or introduced at the node.
+  var scope: Scope? {
+    switch self.syntaxNodeType {
+    case is SourceFileSyntax.Type:
+      FileScope(syntax: self)
+    default:
+      parent?.scope
+    }
+  }
+}
+
+protocol Scope {
+  var parent: Scope? { get }
+
+  var sourceSyntax: SyntaxProtocol { get }
+
+  func getDeclarationFor(name: String, at syntax: SyntaxProtocol) -> Syntax?
+}
+
+extension Scope {
+  var parent: Scope? {
+    getParentScope(forSyntax: sourceSyntax)
+  }
```

Comment:
Yeah, I think this is a good point. As I now think about it, a great part of scopes will have a different parent behavior (like possibly `GenericParameterScope`). I made it a requirement to implement `parent` by all subclasses, but I will also leave `getParentScope` method so that subclasses can still use the default behavior.